### ISSUE TITLE
Clear immutable collection only when condition is met

### ISSUE THREAD

**Issue by cristian-aimi** [mapstruct/mapstruct#3495]
### Expected behavior

```java
if ( h5s.getPurchasedProducts() != null ) {
  if ( savePurchasedProducts( proto.getPurchasedProductsList() ) ) {
      h5s.getPurchasedProducts().clear();
      h5s.getPurchasedProducts().addAll( purchasedProductInputProtoListToPurchasedProductList( proto.getPurchasedProductsList() ) );
  }
}
```

### Actual behavior

```java
if ( h5s.getPurchasedProducts() != null ) {
  h5s.getPurchasedProducts().clear();
  if ( savePurchasedProducts( proto.getPurchasedProductsList() ) ) {
      h5s.getPurchasedProducts().addAll( purchasedProductInputProtoListToPurchasedProductList( proto.getPurchasedProductsList() ) );
  }
}
```

### Steps to reproduce the problem

in the generated code the destination collection is always emptied even if the mapping condition is false, I would expect it not to be touched as happens for the other fields.

### MapStruct Version

1.5.5.Final

**Comment by thunderhook** [comment]
Just to clarify, is `savePurchasedProducts(...)` a `@Condition`? To make it easier for us to reproduce.

**Comment by thunderhook** [comment]
Please provide more code. The mapper class, as well as the source and target types.

I was not able to reproduce this with the following mapper:

```java
package org.mapstruct.ap.test.bugs._3495;

import java.util.List;

import org.mapstruct.Condition;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.MappingTarget;
import org.mapstruct.Named;

/**
 * @author Oliver Erhart
 */
@Mapper
public interface Issue3495Mapper {

    @Mapping(target = "names", source = "names", conditionQualifiedByName = "alwaysFalse")
    void update(Source source, @MappingTarget Target target);

    @Condition
    @Named("alwaysFalse")
    default boolean alwaysFalse(@MappingTarget Target target) {
        return false;
    }

    class Source {

        List<String> names;

        public List<String> getNames() {
            return names;
        }

        public void setNames(List<String> names) {
            this.names = names;
        }
    }

    class Target {

        List<String> names;

        public List<String> getNames() {
            return names;
        }

        public void setNames(List<String> names) {
            this.names = names;
        }

    }
}
```

The generated class looks like:
```java
public class Issue3495MapperImpl implements Issue3495Mapper {

    @Override
    public void update(Source source, Target target) {
        if ( source == null ) {
            return;
        }

        if ( target.getNames() != null ) {
            if ( alwaysFalse( target ) ) {
                target.getNames().clear();
                target.getNames().addAll( source.getNames() );
            }
        }
        else {
            if ( alwaysFalse( target ) ) {
                List<String> list = source.getNames();
                target.setNames( new ArrayList<String>( list ) );
            }
        }
    }
}
```

**Comment by cristian-aimi** [comment]
I checked how your test differs from mine, my target object does not have the "set" method for the List, if you delete the "setNames" method from the target class you should be able to reproduce it.
Now that I understand what the problem is, I added the set method and I've solved it, but the behavior is still dubious.

Thanks so much

**Comment by thunderhook** [comment]
That's a good hint about the missing setter. This is a valid use case. I will try to reproduce it again without the setter and perhaps reopen the issue (when I find time).

**Comment by thunderhook** [comment]
Yes, with an immutable target collection, the target collection will be cleared no matter what. Here is the reproducal:

```java
@Mapper
public interface Issue3495Mapper {

    @Mapping(target = "names", source = "names", conditionQualifiedByName = "alwaysFalse")
    void update(Source source, @MappingTarget Target target);

    @Condition
    @Named("alwaysFalse")
    default boolean alwaysFalse(@MappingTarget Target target) {
        return false;
    }

    class Source {

        private List<String> names;

        public List<String> getNames() {
            return names;
        }

        public void setNames(List<String> names) {
            this.names = names;
        }
    }

    class Target {

        private List<String> names;

        public List<String> getNames() {
            return names;
        }

    }
}
```

```java
// GENERATED CODE
public class Issue3495MapperImpl implements Issue3495Mapper {

    @Override
    public void update(Source source, Target target) {
        if ( source == null ) {
            return;
        }

        if ( target.getNames() != null ) {
            target.getNames().clear();
            if ( alwaysFalse( target ) ) {
                target.getNames().addAll( source.getNames() );
            }
        }
    }
}
```

**Comment by filiphr** [comment]
I've added a comment on the PR (https://github.com/mapstruct/mapstruct/pull/3506#issuecomment-1913667574). I don't believe that the behaviour that is happening when updating a collection without a setter is wrong. It is actually correct. When updating the target collection with some source value you want the target to represent the source. This means that target needs to be cleared and only then it needs to be populated with the data from the source collection.

For updates `NullValuePropertyMappingStrategy` should be used to control what should be happening if source is not set.

**Comment by thunderhook** [comment]
@filiphr What you say sounds logical. It's just confusing because the code behaves differently as soon as a setter is added. Maybe there is something illogical hidden here that I can't grasp right now. 

I will close this issue and the PR.

**Comment by panas-sydor-ei** [comment]
@filiphr, yes, but I assume it shouldn't be like this when the `nullValuePropertyMappingStrategy` is set to `IGNORE`. For example, we have the following case: 

```java
@Mapper(uses = {TittleMapper.class, CollectionsSupportMapper.class}, nullValuePropertyMappingStrategy = IGNORE)
public abstract class PassengerMapper {

  // some other mappings
  // ...

  @Mapping(target = "nameTitles", source = "title", qualifiedByName = "toListOrNull")
  protected abstract void mapPassengerName(final @MappingTarget PersonNameType personName, final Name name);
}
```

The generated code:

```java
    @Override
    protected void mapPassengerName(PersonNameType personName, Name name) {
        if ( name == null ) {
            return;
        }

        if ( personName.getNameTitles() != null ) {
            personName.getNameTitles().clear();
            List<String> list = collectionsSupportMapper.toListOrNull( tittleMapper.toOta( name.getTitle() ) );
            if ( list != null ) {
                personName.getNameTitles().addAll( list );
            }
        }
    }
```

While I would expect it to be generated like the following with `nullValuePropertyMappingStrategy = IGNORE` set:

```java
    @Override
    protected void mapPassengerName(PersonNameType personName, Name name) {
        if ( name == null ) {
            return;
        }

        if ( personName.getNameTitles() != null ) {
            List<String> list = collectionsSupportMapper.toListOrNull( tittleMapper.toOta( name.getTitle() ) );
            if ( list != null ) {
                personName.getNameTitles().clear();
                personName.getNameTitles().addAll( list );
            }
        }
    }
```

**So the target collection is only updated when the source is not null.** 


Also, the same happens when we have a mapping with a condition:
```java
@Mapping(target = "nameTitles", source = "title", qualifiedByName = "toListOrNull", conditionQualifiedByName = "isNotNull")
```
The generated code:
```java
    @Override
    protected void mapPassengerName(PersonNameType personName, Name name) {
        if ( name == null ) {
            return;
        }

        if ( personName.getNameTitles() != null ) {
            personName.getNameTitles().clear();
            if ( collectionsSupportMapper.isNotNull( name.getTitle() ) ) {
                personName.getNameTitles().addAll( collectionsSupportMapper.toListOrNull( tittleMapper.toOta( name.getTitle() ) ) );
            }
        }
    }
```

**Comment by filiphr** [comment]
I think that you are right @panas-sydor-ei. I believe that we actually resolved this with #3806. The user in that issue managed to provide an example where we were actually a bit inconsistent between our different implementations
